chalk/core.py

```diff
@@ -198,47 +204,44 @@ def _repr_svg_(self) -> str:
         os.unlink(f.name)
         return svg
 
-    def _empty_check(
-        self, other: Diagram
-    ) -> Tuple[Optional[Diagram], BoundingBox, BoundingBox]:
-        fake = BoundingBox([Point(0, 0)])
-        box1 = self.get_bounding_box()
-        box2 = other.get_bounding_box()
-        if box2 is None:
-            return self, fake, fake
-        if box1 is None:
-            return other, fake, fake
-        return None, box1, box2
+    def with_envelope(self, other: Diagram) -> Diagram:
+        return Compose(other.get_envelope(), self, Empty())
+
+    def juxtapose(self, other: Diagram, direction: Vec2) -> Diagram:
+        """Given two diagrams ``a`` and ``b``, ``a.juxtapose(b, v)``
+        places ``b`` to touch ``a`` along angle ve .
+
+        Args:
+            other (Diagram): Another diagram object.
+            direction (Vec2): (Normalized) vector angle to juxtapose
+
+        Returns:
+            Diagram: Repositioned ``b`` diagram
+        """
+        envelope1 = self.get_envelope()
+        envelope2 = other.get_envelope()
+        d = envelope1.envelope_v(direction) - envelope2.envelope_v(-direction)
+        t = Affine.translation(d)
+        return ApplyTransform(t, other)
 
     def atop(self, other: Diagram) -> Diagram:
-        dia, box1, box2 = self._empty_check(other)
-        if dia is not None:
-            return dia
-        new_box = BoundingBox.from_shapes([box1, box2])
-        return Compose(new_box, self, other)
+        envelope1 = self.get_envelope()
+        envelope2 = other.get_envelope()
+        new_envelope = envelope1 + envelope2
+        return Compose(new_envelope, self, other)
 
     __add__ = atop
 
-    def _merge(self, other: Diagram, direction: Vec2) -> Diagram:
-        dia, box1, box2 = self._empty_check(other)
-        if dia is not None:
-            return dia
-        d = box1.max_point - box2.min_point
-        t = Affine.translation(direction * d)
-        new_box = BoundingBox.from_shapes(
-            [b for b in [box1, t * box2] if b is not None]
-        )
-        return Compose(new_box, self, ApplyTransform(t, other))
-
     def above(self, other: Diagram) -> Diagram:
-        return self._merge(other, unit_y)
+        return self.beside(other, unit_y)
 
     __truediv__ = above
 
-    def beside(self, other: Diagram) -> Diagram:
-        return self._merge(other, unit_x)
+    def beside(self, other: Diagram, direction: Optional[Vec2]) -> Diagram:
```

Comment:
From my reading, I think that the type of `direction` should be `Vec2`, as `juxtapose` doesn't know what to do with a `None`.